### PR TITLE
[4.1] Invalid caching of WebAssets

### PR DESCRIPTION
**Pull request by Denitz**
### Summary of Changes

1. `Cache::setWorkarounds()` incorrectly returns ALL current assets even if `headerbefore` option exists and a module/view didn't actually add any own assets.

This results in incremental caching of all assets in the cache of each module. Tons of WebAssets events are next executed on module cache load via `Cache::getWorkarounds()` and the load time is seriously affected.

2. Iterating over `$wa->getManagerState()['activeAssets']` does not requires checking the `ASSET_STATE_ACTIVE` state because this array holds only active assets

3. The resulting array of `getHeadData()['assetManager']['assets']` didn't have the asset name as key which prevents the fast cache workarounds comparison

### Testing Instructions

Enable Joomla conservative caching, browse frontend pages as guest

### Actual result BEFORE applying this Pull Request

Explore the contents of module cache files, see that they contain all `WebAssetItem` objects existing at the time of caching.
Note the cache file size.

### Expected result AFTER applying this Pull Request

See that cache file size in much smaller now (delete old files first), see that the module cache now only have the assets really added by this module.

### Documentation Changes Required

No.

### CODE REVIEW

File: libraries/src/Document/HtmlDocument.php [hunk 1]
```diff
@@ -172,10 +172,7 @@ public function getHeadData()
 		{
 			foreach ($assetNames as $assetName => $assetState)
 			{
-				if ($assetState === WebAssetManager::ASSET_STATE_ACTIVE)
-				{
-					$waState['assets'][$assetType][] = $wa->getAsset($assetType, $assetName);
-				}
+				$waState['assets'][$assetType][$assetName] = $wa->getAsset($assetType, $assetName);
```
Comment (Fedik) [inline libraries/src/Document/HtmlDocument.php:176]: Please revert it, this was made to skip assets with `WebAssetManager::ASSET_STATE_DEPENDENCY`
to reduce amount of cached data.

There a comment above, at line 167.

Comment (Fedik) [reply libraries/src/Document/HtmlDocument.php:176]: hmhm, if dependency registered at runtime, then better keep it.
well, ignore my previous comment :smile:

**Review by Fedik — APPROVED** [review]
looks good to me

File: libraries/src/Document/HtmlDocument.php (hunk 1 above)
Comment (Denitz) [reply libraries/src/Document/HtmlDocument.php:176]: `$waState['activeAssets']` holds ONLY assets with `WebAssetManager::ASSET_STATE_ACTIVE` status

Comment (Fedik) [reply libraries/src/Document/HtmlDocument.php:176]: > holds ONLY assets ASSET_STATE_ACTIVE

nope, not always,
but nevermind

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 2c1296222529c01cd260d3e6d77ce5119ccfcdc2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35776">issues.joomla.org/tracker/joomla-cms/35776</a>.</sub>

**Comment by Denitz** [comment]
Please check https://github.com/joomla/joomla-cms/pull/37139 with better approach.

**Comment by richard67** [comment]
> Please check #37139 with better approach.

@Denitz Does that PR replace this one here? If so, this one here should be closed.

**Pull request closed without merging**